input/Glfw/GlfwInputDevice.cpp

```diff
@@ -351,6 +367,139 @@ namespace NovelRT::Input::Glfw
+    {
+        _mousePos = DetermineMouseScreenPosition(pos);
+    }
+
+    void GlfwInputDevice::ProcessKeyState(int32_t key, KeyState state)
+    {
+        auto& previousBuffer = _keyStates.at(_previousBufferIndex);
+        auto& currentBuffer = _keyStates.at(_currentBufferIndex);
+
+        KeyState previousStateResult = KeyState::Idle;
+
+        for (auto& [previousKey, previousLog] : previousBuffer)
+        {
+            if (previousKey == key)
+            {
+                previousStateResult = previousLog.GetCurrentState();
+                break;
+            }
+        }
+
+        KeyStateFrameChangeLog changeLogObject = KeyStateFrameChangeLog();
+
+        for (auto& [currentKey, currentLog] : currentBuffer)
+        {
+            if (currentKey == key)
+            {
+                changeLogObject = currentLog;
+                break;
+            }
+        }
+
+        switch (state)
+        {
+            case KeyState::KeyDown:
+                if (previousStateResult == KeyState::KeyDown)
+                {
+                    changeLogObject.PushNewState(KeyState::KeyDownHeld);
+                }
+                else if (previousStateResult != KeyState::KeyDownHeld)
+                {
+                    changeLogObject.PushNewState(KeyState::KeyDown);
+                }
+                break;
+            case KeyState::KeyDownHeld:
+            case KeyState::KeyUp:
+                changeLogObject.PushNewState((previousStateResult == KeyState::KeyUp) ? KeyState::Idle : state);
+                break;
+            case KeyState::Idle:
+            default:
+                break;
+        }
+
+        currentBuffer.insert_or_assign(key, changeLogObject);
+    }
+
+    NovelRT::Maths::GeoVector2F GlfwInputDevice::DetermineMouseScreenPosition(NovelRT::Maths::GeoVector2F& pos)
+    {
+        return NovelRT::Maths::GeoVector2F(static_cast<float>(pos.x - (_windowDimensions.x / 2)),
+                                           static_cast<float>(-pos.y + (_windowDimensions.y / 2)));
+    }
+
+    KeyStateFrameChangeLog GlfwInputDevice::GetCurrentChangeLog(const std::string& key)
+    {
+        for (auto& action : _mappedActions)
+        {
+            if (action.actionName != key)
+                continue;
+
+            return _keyStates.at(_currentBufferIndex).at(action.pairedKey.GetExternalKeyCode());
+        }
+
+        return KeyStateFrameChangeLog();
+    }
+
+    KeyStateFrameChangeLog GlfwInputDevice::GetPreviousChangeLog(const std::string& key)
+    {
+        for (auto& action : _mappedActions)
+        {
+            if (action.actionName != key)
+                continue;
+
+            return _keyStates.at(_previousBufferIndex).at(action.pairedKey.GetExternalKeyCode());
+        }
+
+        return KeyStateFrameChangeLog();
+    }
+
+    NovelRT::Utilities::Misc::Span<std::unordered_map<int32_t, KeyStateFrameChangeLog>> GlfwInputDevice::GetAllChangeLogs()
```

Comment:
Is there a reason this method is mutable? Just curious.

Reply:
Answered your own question from next review statement (lol). Resolving.